Feature/incoming webhooks

titpetric opened the pull request:
Initial webhooks implementation:

- [x] needs routing configuration with/without jwt middleware,
- [x] needs API sanity check (editing webhooks),
- [x] create a bot user for a webhook message
- [x] manage webhook user avatars

darh commented on api/messaging/spec/webhooks.json:
```diff
@@ -0,0 +1,107 @@
+            "type": "uint64"
+          },
+          {
+            "name": "webhookToken",
```
should this be at least GET if not header?

darh replied:
I suggest we (re)use JWT here, no need for additional authentication here.

titpetric replied:
Agreed. JWT should be used on API routes for validation. Effectively it's going to be a 100 year expiry, so we'll still need to validate it against the DB. Will resolve comment thread when done.

titpetric replied:
I opted against this, since:

1. the api route doesn't need middleware validation (exclusive route),
2. we don't want to leak the jwt identity claims for a webhook to other APIs,
3. that leaves the only jwt feature that would be used: expiry (which we don't need)

Effectively any random string/uuid will do.

darh commented on api/messaging/spec/webhooks.json:
```diff
@@ -0,0 +1,107 @@
+      "Name": "webhookGet",
+      "Method": "GET",
+      "Title": "Get webhook details",
+      "Path": "/webhook/{webhookID}",
```
`/webhooks/webhook/{webhookID}` 
one level too deep?

darh commented on messaging/internal/service/channel.go:
```diff
@@ -54,6 +55,9 @@ type (
 		Undelete(ID uint64) (*types.Channel, error)
 		RecordView(userID, channelID, lastMessageID uint64) error
+
+		WebhookList(channelID uint64) (types.WebhookSet, error)
```
let's reverse this --- channels should not be aware of webhooks, webhooks are always (?) aware of channels.

darh commented on messaging/internal/service/webhook.go:
```diff
@@ -0,0 +1,81 @@
+		Delete(webhookID uint64) error
+		DeleteWithToken(webhookID uint64, webhookToken string) error
+
+		PublishMessage(webhookID uint64, webhookToken string, username, avatar, message string) (interface{}, error)
```
rename publish => create

titpetric commented:
update:

- add the concept of an outgoing webhook,
- restructure/fix specs and functions on denis's comments

titpetric commented:
Important changes introduced:

- tests split into unit / integration and external (documented in wiki),
- permissions have an allow-everything mode for tests via context.WithValue (still requires identity to pass),
- test based entrypoints in individual `app/service`, needs *testing.T to get full API access
- full integration tests suite, test fixes
- permissions for webhooks as discussed,
- updated static assets packager to goware/statik,
- moved mocks into subfolders (service/mocks/*),
- add commands that trigger outgoing webhooks as the ultimate fallback,
- various style refactoring, constructors, naming, etc.

titpetric commented:
This is a functional webhooks implementation now. Future improvements:

- actually implement avatar storage (should be shared between messaging and system),
- fix webapp* to honor meta fields on systemTypes.User (avatar) and messaging.Message (username, avatar)

titpetric commented on api/messaging/spec/webhooks.json:
```diff
@@ -0,0 +1,183 @@
+  "Path": "/webhooks",
+  "APIs": [
+    {
+      "Name": "webhookList",
```
done

titpetric commented on api/messaging/spec/webhooks_public.json:
```diff
@@ -0,0 +1,75 @@
+      "Name": "webhookDelete",
+      "Method": "DELETE",
+      "Title": "Delete webhook",
+      "Path": "/{webhookID}/{webhookToken}",
```
It resolves a possible security issue where a webhook may be exposed, and a user with permissions to delete it might be unavailable - but the webhook URL is accessible. 

> - the reasoning behing this one is more out of convenience. If some webhook link would leak and would be discovered outside of the sysadmin teams, they could invalidate the webhook just by issuing a DELETE statement. Implementation at this point: optional, but easy.

Documented as motivation in #101

titpetric commented on crm/internal/service/permissions_test.go:
```diff
@@ -39,33 +44,31 @@ func TestPermissions(t *testing.T) {
-	permissionsSvc := Permissions().With(ctx)
-	systemRulesSvc := systemService.Rules().With(ctx)
+	cleanContext := auth.SetIdentityToContext(context.Background(), user)
+	permissionsSvc := Permissions().With(cleanContext)
```
It is, we're testing the permission svc itself.

titpetric commented on crm/internal/service/record_test.go:
```diff
@@ -7,28 +9,31 @@ import (
 	{
 		err := user.GeneratePassword("Mary had a little lamb, little lamb, little lamb")
-		assert(t, err == nil, "Error generating password: %+v", err)
+		test.Assert(t, err == nil, "Error generating password: %+v", err)
```
it's a simple question but the answer isn't :)

- userSvc Create might require a password to insert user,
- GeneratePassword returns error (we can't asume it won't),
- just always check errors, the above check is effectively free

darh commented on system/internal/service/user.go:
```diff
@@ -163,14 +171,39 @@ func (svc *user) Update(mod *types.User) (u *types.User, err error) {
 	})
 }
 
+func (svc *user) bindAvatar(user *types.User, avatar *multipart.FileHeader, avatarURL string) (*types.User, error) {
```
this is not ok, we should not be dependant on http inside service layer. 
file should be resolved inside controller and service should get io.Reader (or a similar one from that family)

darh commented on messaging/internal/service/webhook.go:
```diff
@@ -0,0 +1,258 @@
+
+		users   systemService.UserService
+		webhook repository.WebhookRepository
+		message repository.MessageRepository
```
Using Message repository directly will result in circumventing message creation logic that service covers (validation, mentions, unread inc, notifying subscribers....)

This must be changed.

titpetric commented on messaging/internal/service/channel.go:
```diff
@@ -665,10 +663,11 @@ func (svc *channel) AddMember(channelID uint64, memberIDs ...uint64) (out types.
-				return errors.New("not allowed to join")
-			} else if !svc.prm.CanManageChannelMembers(ch) {
-				return errors.New("not allowed to add channel members")
+			if memberID == userID && !svc.perms.CanJoinChannel(ch) {
```
Just to clarify, similar checks were only broken down from if/else chains if the previous if had a finalizer (return) statement. Go vet or metalinter or critic reports this as a refactor job from if/else into a switch.
